Hey, @arshajii!

Before pushing this PR, I also thought this issue might be related with GV un-init problem.

After implement done, I checked ptx of #781 MRE, and this is what it looks like:

```
//
// Generated by LLVM NVPTX Back-End
//

.version 4.2
.target sm_30
.address_size 64

	// .globl	kernel_0_0_std_internal_types_array_List_0_float__
.visible .global .align 4 .f32 _pi32_0;

.visible .entry kernel_0_0_std_internal_types_array_List_0_float__(
	.param .u64 .ptr .align 1 kernel_0_0_std_internal_types_array_List_0_float___param_0
)
{
	.reg .b32 	%r<2>;
	.reg .f32 	%f<2>;
	.reg .b64 	%rd<6>;
	.reg .f64 	%fd<2>;

	ld.param.u64 	%rd1, [kernel_0_0_std_internal_types_array_List_0_float___param_0];
	cvta.to.global.u64 	%rd2, %rd1;
	mov.u32 	%r1, %tid.x;
	ld.global.f32 	%f1, [_pi32_0];
	cvt.f64.f32 	%fd1, %f1;
	ld.global.u64 	%rd3, [%rd2+16];
	mul.wide.u32 	%rd4, %r1, 8;
	add.s64 	%rd5, %rd3, %rd4;
	st.f64 	[%rd5], %fd1;
	ret;

}

```

## result

```
$ ./pi32_gpu_f32
out:  0
math.pi32:  3.14159
AssertionError: Assert failed (/home/swchoi/src/test_code/codon_math_matrix/generated/cases/nvptx/pi32_gpu_f32.codon:17)

Raised from: 
/home/swchoi/src/test_code/codon_math_matrix/generated/cases/nvptx/pi32_gpu_f32.codon:17:1
Aborted (core dumped)
```

Result seems to say that it still un-inits GV.

--- a/codon/cir/llvm/gpu.cpp
+++ b/codon/cir/llvm/gpu.cpp
@@ -945,7 +945,7 @@ void patchPTXVar(llvm::Module *M, llvm::GlobalValue *ptxVar,
 }
 } // namespace
 
-std::unique_ptr<llvm::Module> prepareGPUmodule(llvm::Module *M){
+std::unique_ptr<llvm::Module> prepareGPUmodule(llvm::Module *M) {
   std::unique_ptr<llvm::Module> clone = llvm::CloneModule(*M);
   clone->setTargetTriple(llvm::Triple::normalize(GPU_TRIPLE));
   clone->setDataLayout(GPU_DL);
@@ -956,8 +956,9 @@ std::unique_ptr<llvm::Module> prepareGPUmodule(llvm::Module *M){
   return clone;
 }
 
-void applyGPUTransformations(llvm::Module *M, std::unique_ptr<llvm::Module> clone, const std::string &ptxFilename) {
-    llvm::LLVMContext &context = M->getContext();
+void applyGPUTransformations(llvm::Module *M, std::unique_ptr<llvm::Module> clone,
+                             const std::string &ptxFilename) {
+  llvm::LLVMContext &context = M->getContext();
   llvm::NamedMDNode *nvvmAnno = clone->getOrInsertNamedMetadata("nvvm.annotations");
   std::vector<llvm::Function *> kernelCandidates;
   std::vector<llvm::GlobalValue *> kernels;

--- a/codon/cir/llvm/gpu.h
+++ b/codon/cir/llvm/gpu.h
@@ -9,15 +9,21 @@
 namespace codon {
 namespace ir {
 
-/// Applies GPU-specific transformations and generates PTX
-/// code from kernel functions in the given LLVM module.
-/// @param module LLVM module containing GPU kernel functions (marked with "kernel"
-/// annotation)
-/// @param ptxFilename Filename for output PTX code; empty to use filename based on
-/// module
+/// Applies GPU-specific transformations to the prepared GPU module, generates
+/// PTX for kernel functions, and patches the original module accordingly.
+/// @param module Original LLVM module to patch with generated GPU artifacts.
+/// @param clone GPU-targeted module produced by prepareGPUmodule(); ownership is
+/// consumed.
+/// @param ptxFilename Filename for output PTX code; empty to derive it from the module.
 
+void applyGPUTransformations(llvm::Module *module, std::unique_ptr<llvm::Module> clone,
+                             const std::string &ptxFilename = "");
+
+/// Creates a GPU-targeted clone of the given LLVM module.
+/// The cloned module is configured with the GPU target triple and data layout
+/// so it can go through the GPU-specific optimization pipeline.
+/// @param module LLVM module containing both host code and GPU kernel functions.
+/// @return A cloned module configured for GPU lowering.
 std::unique_ptr<llvm::Module> prepareGPUmodule(llvm::Module *module);
-void applyGPUTransformations(llvm::Module *module, std::unique_ptr<llvm::Module> clone, const std::string &ptxFilename = "");
-
 } // namespace ir
 } // namespace codon

--- a/codon/cir/llvm/optimize.cpp
+++ b/codon/cir/llvm/optimize.cpp
@@ -1036,9 +1036,9 @@ void runLLVMOptimizationPasses(llvm::Module *module, bool debug, bool jit,
   llvm::FunctionAnalysisManager fam;
   llvm::CGSCCAnalysisManager cgam;
   llvm::ModuleAnalysisManager mam;
-  auto machine =
-      includeNative ? getTargetMachine(module, /*setFunctionAttributes=*/true)
-                    : std::unique_ptr<llvm::TargetMachine>();
+  auto machine = includeNative
+                     ? getTargetMachine(module, /*setFunctionAttributes=*/true)
+                     : std::unique_ptr<llvm::TargetMachine>();
   llvm::PassBuilder pb(machine.get());
 
   llvm::Triple moduleTriple(module->getTargetTriple());
@@ -1091,11 +1091,15 @@ void optimize(llvm::Module *module, bool debug, bool jit, PluginManager *plugins
   }
   {
     TIME("llvm/opt1");
-    runLLVMOptimizationPasses(module, debug, jit, plugins, true, true);
+    runLLVMOptimizationPasses(module, debug, jit, plugins,
+                              /*includeNative=*/true,
+                              /*includePlugins=*/true);
   }
   if (!debug) {
     TIME("llvm/opt2");
-    runLLVMOptimizationPasses(module, debug, jit, plugins, true, true);
+    runLLVMOptimizationPasses(module, debug, jit, plugins,
+                              /*includeNative=*/true,
+                              /*includePlugins=*/true);
   }
   {
     TIME("llvm/gpuopt1");

